### DISCUSSION TITLE
Can regex support mismatching '\n'?

### DISCUSSION THREAD

**Discussion by freekeeper** [rust-lang/regex#920] [Q&A] [answered]
[RE2](https://github.com/google/re2) support does not match newlines. Like [this](https://github.com/google/re2/blob/main/re2/testing/re2_test.cc#L1320-L1346), by setting `dot_nl_ = true`, the regular expression can not match the `\n` character.

Does [regex](https://github.com/rust-lang/regex) have such a function, or in what way can it be achieved?

**Comment by BurntSushi** [comment, marked as answer]
`.` does not match `\n` by default, like just about every other regex engine. And like almost every other regex engine, you can make `.` match `\n` by enabling the `s` flag.

**Reply by freekeeper** [reply]
There are the following test cases in [RE2](https://github.com/google/re2/blob/main/re2/testing/re2_test.cc#L1320-L1346)
```
static struct NeverTest {
  const char* regexp;
  const char* text;
  const char* match;
} never_tests[] = {
  { "(.*)", "abc\ndef\nghi\n", "abc" },
  { "(?s)(abc.*def)", "abc\ndef\n", NULL },
  { "(abc(.|\n)*def)", "abc\ndef\n", NULL },
  { "(abc[^x]*def)", "abc\ndef\n", NULL },
  { "(abc[^x]*def)", "abczzzdef\ndef\n", "abczzzdef" },
};

TEST(RE2, NeverNewline) {
  RE2::Options opt;
  opt.set_never_nl(true);
  for (size_t i = 0; i < arraysize(never_tests); i++) {
    const NeverTest& t = never_tests[i];
    RE2 re(t.regexp, opt);
    if (t.match == NULL) {
      EXPECT_FALSE(re.PartialMatch(t.text, re));
    } else {
      StringPiece m;
      EXPECT_TRUE(re.PartialMatch(t.text, re, &m));
      EXPECT_EQ(m, t.match);
    }
  }
}
```
In the above test case, `never_tests[1].regexp = "(?s)(abc.*def)"`, `never_tests[1].text = "abc\ndef\n"`, `(?s)` means make `.` match `\n`. 

But by setting `opt.set_never_nl(true)`, all regular expressions cannot match `\n`. Can this pattern be implemented in `regex` or in what way?

According to my reading of the `regex` source code, I found that `regex` implements the [dot_nl](https://github.com/rust-lang/regex/blob/master/regex-capi/include/rure.h#L48) mode in [RE2](https://github.com/google/re2/blob/main/re2/re2.h#L736). But `regex` does not seem to implement the `never_nl` mode in [RE2](https://github.com/google/re2/blob/main/re2/re2.h#L736). So I wonder if `regex` can implement the `never_nl` mode in `RE2`.

**Reply by BurntSushi** [reply]
`never_nl` and `dot_nl` are two very different options. There is no out-of-the-box `never_nl` option in this crate, but you can implement it without too much fuss yourself by using the `regex-syntax` crate and transforming the `Hir`. Here's the code ripgrep uses to do that exact thing: https://github.com/BurntSushi/ripgrep/blob/65b1b0e38a973ce15d887a1b01d27cabe391cd24/crates/regex/src/strip.rs

**Reply by freekeeper** [reply]
Thank you so much!
